전남대 BE_박규현 2주차 과제(step2~step3)

pinkbeenyeti opened the pull request:
코드 리뷰 수정 완료하였습니다!


어려웠던 점
1. 토큰의 인증으로 사용자를 구분하는 것이 가장 어려웠던 것 같습니다. 굉장히 많은 시도를 하였는데, 계속해서 Authrization 필드를 찾지 못했다는 오류가 떴습니다.. 어떤 곳에서 발생한 것인 지 궁금합니다.

pinkbeenyeti committed:
fix: 피드백 반영 post 반환 수정
기존에 상품에 대한 값을 수정 시 서버에서 responseEntity를 반환하였으나
responseEntity를 반환하지 않고 String을 반환하여 원래의 패이지로 바로 돌아갈
수 있도록 수정

pinkbeenyeti committed:
fix: 오루 수정
필요하지 않은 부분들이 추가된 부붑을 수정

pinkbeenyeti committed:
feat: schema.sql 파일 생성
스키마 스크립트를 생성하기 위한 파일을 생성하였다.

pinkbeenyeti committed:
feat: propertiy 업데이트
스키마 스키립트를 사용하기 위해 속성을 추가하였다.

pinkbeenyeti committed:
fix: 스키마 스크립트를 사용하도록 Application 클래스 수정

pinkbeenyeti committed:
fix : addProduct, 모달 창을 담당하는 코드 수정
post의 수행을 check.js가 하도록 post 하는 부분을 삭제하였다.

pinkbeenyeti committed:
fix: editProduct 수정
사용자가 product의 정보 수정 시 id를 변경하지 않도록 하기 위해 아이디를
변경하는 부분을 삭제하였다.

pinkbeenyeti committed:
feat: 예외 검출, validation 이용
Validation을 이용하여 객체인 ProductDTO의 필드인 name에서 길이 제한과
정규식을 이용한 제한을 두어 예외를 던지도록 설정하였다.

pinkbeenyeti committed:
feat: 예외 객체
카카오라는 문구가 들어갈 경우 예외를 던지기 위해 만든 객체

pinkbeenyeti committed:
feat: 전역에서 발생한 예외를 처리하는 컨트롤러
글자 수, 특수문자, 카카오 문구 포함 시 발생하는 예외를 검출하고
응답한다. 해당 예외 발생 시 400 상태 코드와 에러 메시지를 클라이언트에
전송한다.

pinkbeenyeti committed:
fix: 자바 스크립트 수정
기존의 코드에서 400 상태 코드를 받을 경우에만 해당 에러 메시지를
출력하도록 수정한다.

pinkbeenyeti committed:
feat: 컨트롤러가 유효성을 검사할 수 있도록 추가
컨트롤러가 유효성을 검사할 수 있도록 세팅한다.
추가로 name이 카카오라는 문구를 포함할 경우 예외를 던지도록 한다.

pinkbeenyeti committed:
fix: controller에서 simpleJdbcInsert를 사용하도록 변경
simpleJdbcInsert를 사용하여 코드의 가독성을 증가

pinkbeenyeti committed:
fix: 컨트롤러 메서드 분리
sql문과 카카오 예외를 던지는 행위가 반복 되므로 메서드를 분리한 후
사용도록 하여 코드의 가독성을 증가시켰다.

pinkbeenyeti committed:
feat: readme 작성

pinkbeenyeti committed:
fix: 파일 이름 수정
각 코드가 작성된 파일의 기능을 명확하게 하기 위해 파일 이름을 수정함

pinkbeenyeti committed:
feat: 로그인 화면 구현
간다한 로그인을 할 수 이는 웹사이트 뷰를 제작

pinkbeenyeti committed:
fix: 로그인 화면의 password 영역 id 이름 수정

pinkbeenyeti committed:
feat: 로그인 화면의 스타일 생성

pinkbeenyeti committed:
feat: 로그인 화면의 js 생성
사용자가 입력한 이메일과 password를 post 메소드로 전송한다.
응답으로 200이 오지 않는 다면 alert 창에 전송받은 경고 문구를 출력한다.
응답이 200이라면 전송 받은 토큰을 alert에 띄운다

pinkbeenyeti committed:
feat: 회원가입 화면 구현

pinkbeenyeti committed:
feat: 회원가입 화면의 디자인 구현
register.css 구현

pinkbeenyeti committed:
feat: 회원가입 화면의 기능 구현
submit 버튼을 통해 사용자가 입력한 email과 password 제출
응답을 받아 상태코드 200이라면 전송받은 토큰 출력
200이 아니라면 전송받은 오류메시지 출력

pinkbeenyeti committed:
feat: 의존성 추가
토큰을 생성하기 위해 jjwt를 사용하기 위한 의존성 추가

pinkbeenyeti committed:
feat: 사용자의 이메일과 패스워드를 저장하는 테이블 추가

pinkbeenyeti committed:
feat: 로그인과 회원가입 페이지에서 사용자의 요청을 받는 각각의 컨트롤러
생성

pinkbeenyeti committed:
feat:계층 간의 데이터를 주고 받기 위한 dto 생성

pinkbeenyeti committed:
feat: 회원가입 페이지의 응답을 받는 컨트롤러 추가

pinkbeenyeti committed:
feat: 계층 및 클라이언트와 토큰을 주고 받기 위한 dto

pinkbeenyeti committed:
feat: 각 페이지에서의 비즈니스 로직을 처리하는 계층

pinkbeenyeti committed:
feat: users db와 실제 상호작용하는 계층

pinkbeenyeti committed:
feat: users DB와 실제 상호작용을 하기 위한 객체

pinkbeenyeti committed:
fix: 약간의 가독성 개선

pinkbeenyeti committed:
feat: readme.md 작성

dc-choi commented on src/main/java/gift/controller/LoginController.java:
```diff
@@ -0,0 +1,61 @@
+
+@Validated
+@Controller
+@RequestMapping("/login")
```
step2 과제에서 요구하는 URL 주소로 다시 매핑시켜주세요.

pinkbeenyeti replied:
@RequestMapping("/members/login")
위와 같이 완료했습니다!

dc-choi commented on src/main/java/gift/controller/RegisterController.java:
```diff
@@ -0,0 +1,50 @@
+
+@Validated
+@Controller
+@RequestMapping("/register")
```
step2 과제에서 요구하는 URL 주소로 다시 매핑시켜주세요.

pinkbeenyeti replied:
@RequestMapping("/members/register")
위와 같이 완료 했습니다!

dc-choi commented on src/main/java/gift/controller/ProductController.java:
```diff
@@ -0,0 +1,97 @@
+@Controller
+@Validated
+@RequestMapping("/")
+public class ProductController {
```
`Controller` 레이어의 역할에 대해서 고민해주세요. 지금은 모든 로직이 전부 `Controller`에 있습니다.

pinkbeenyeti replied:
user와 마찬가지로 컨트롤러, 레포지토리, 서비스로 분리하여 구현하였습니다!

dc-choi commented on src/main/java/gift/controller/LoginController.java:
```diff
@@ -0,0 +1,61 @@
+    @PostMapping
+    public ResponseEntity<?> loginUser(@Valid @RequestBody UserDTO user) {
+        boolean userExists = userService.redundantUser(user);
+        if (!userExists) {
```
에러처리에 대한 고민이 부족해보입니다. 스프링에서 예외처리를 어떻게 하는가에 대한 고민이 필요해보입니다.

pinkbeenyeti replied:
다음과 같이 변경해 보았습니다!

```
@PostMapping
    public ResponseEntity<?> loginUser(@Valid @RequestBody UserDTO userDTO) {
        authService.redundantUser("login", userDTO);
        authService.comparePassword(userDTO);

        return new ResponseEntity<>(authService.makeToken(userDTO), HttpStatus.OK);
    }
```

```
public void redundantUser(String state, UserDTO userDTO) {
        boolean userExists = userRepository.countUsers(userDTO.email()).orElse(0) > 0;

        if (state.equals("login") & !userExists) {
            throw new InvalidUserException(ErrorMessage.emailNotExists);
        }

        if (state.equals("regist") & userExists) {
            throw new InvalidUserException(ErrorMessage.emailAlreadyExists);
        }
    }
```

dc-choi commented on src/main/java/gift/controller/LoginController.java:
```diff
@@ -0,0 +1,61 @@
+        }
+
+        boolean validLogin = authService.comparePassword(user);
+        if(validLogin) {
```
마찬가지로 에러처리에 대한 고민이 부족해보입니다. 스프링에서 예외처리를 어떻게 하는가에 대한 고민이 필요해보입니다.

pinkbeenyeti replied:
위와 마찬가지로 변경하였습니다!

```
public void comparePassword(UserDTO userDTO) {
        User realUser = userRepository.selectUser(userDTO.email());
        if(!realUser.getPassword().equals(userDTO.password())) {
            throw new InvalidUserException(ErrorMessage.passwordInvalid);
        }
    }
```

dc-choi commented on src/main/java/gift/controller/RegisterController.java:
```diff
@@ -0,0 +1,50 @@
+    public ResponseEntity<?> registUser(@Valid @RequestBody UserDTO user) {
+        boolean userExists = userService.redundantUser(user);
+
+        if (userExists) {
```
마찬가지로 에러처리에 대한 고민이 부족해보입니다. 스프링에서 예외처리를 어떻게 하는가에 대한 고민이 필요해보입니다.

pinkbeenyeti replied:
변경하였습니다!

```
@PostMapping
    public ResponseEntity<?> registUser(@Valid @RequestBody UserDTO userDTO) {
        authService.redundantUser("regist", userDTO);

        userService.createUser(userDTO);
        TokenDTO token = authService.makeToken(userDTO);
        return new ResponseEntity<>(token, HttpStatus.OK);
    }
```

dc-choi commented on src/main/java/gift/dto/ProductDTO.java:
```diff
@@ -1,6 +1,14 @@
-                       Integer price,
-                       String imageUrl) { }
+                         @Size (max=15, message = "이름이 15글자를 초과하였습니다.")
+                         @Pattern(
```
이 부분에서 예외가 발생하면 어떤 예외 객체가 생성되나요?

pinkbeenyeti replied:
MethodArgumentNotValidException이 생성되는 것으로 알고 있습니다!

dc-choi commented on src/main/java/gift/exceptions/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.HashMap;
+import java.util.Map;
+
+@ControllerAdvice
```
`@ControllerAdvice`란 무엇인가요?

pinkbeenyeti replied:
SPRING에서 제공하는 애플리케이션 내 모든 컨트롤러에서 발생하는 예외를 처리하게 해주는 기능이라고 알고 있습니다!

dc-choi commented on src/main/java/gift/exceptions/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.Map;
+
+@ControllerAdvice
+public class GlobalExceptionHandler {
```
`GlobalExceptionHandler`로 모든 예외를 처리한다면 다른 로직에서 이 부분을 호출하기 위해서 어떻게 처리해야하는지에 대한 고민이 필요해보입니다.

pinkbeenyeti replied:
@ControllerAdvice를 통해 SPRING에서 제공하는 애플리케이션 내 모든 컨트롤러에서 발생하는 예외를 처리하게 되므로 컨트롤러에서 발생하는 예외에 대한 처리 메서드를 추가하면 될까요?

dc-choi commented on src/main/java/gift/service/UserService.java:
```diff
@@ -0,0 +1,41 @@
+    public TokenDTO makeToken(UserDTO user) {
+        String secretKey = "Yn2kjibddFAWtnPJ2AFlL8WXmohJMCvigQggaEypa5E=";
+        String accessToken = Jwts.builder()
+                .setSubject(user.email())
```
회원의 권한에 대한 부분이 누락되어 보입니다. 회원의 권한을 생각해주세요.

pinkbeenyeti replied:
```
    public TokenDTO makeToken(UserDTO userDTO) {
        String accessToken = Jwts.builder()
                .setSubject(userDTO.email())
                .claim("role", "user")
                .signWith(Keys.hmacShaKeyFor(secretKey.getBytes()))
                .compact();

        return new TokenDTO(accessToken);
    
```
claim 메서드를 통해 role이라는 키에 user이라는 값을 넣어 회원의 권한이 user에만 있도록 설정했습니다!

dc-choi commented on src/main/java/gift/service/UserService.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    public TokenDTO makeToken(UserDTO user) {
+        String secretKey = "Yn2kjibddFAWtnPJ2AFlL8WXmohJMCvigQggaEypa5E=";
```
`secretKey`는 원칙적으로 노출이 되면 안됩니다. JWT를 생성하고 암호화하는 Key라서 유출되는 경우 보안에 위험합니다. 참고하세요.

pinkbeenyeti replied:
application.properties에 `jwt.secretKey=Yn2kjibddFAWtnPJ2AFlL8WXmohJMCvigQggaEypa5E=`
이라는 속성을 추가하고

```
@Value("${jwt.secretKey}")
 private String secretKey;
```

@Value를 사용하여 appication.properties에 있는 jwt.secretKey에 있는 값을 가져와 변수 secretKey에 할당하는 식으로 구현해 보았습니다!

dc-choi commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,14 @@
+);
+
+DROP TABLE IF EXISTS USERS;
+CREATE TABLE USERS (
```
회원의 권한에 대한 부분이 누락되어 보입니다. 회원의 권한을 생각해주세요.

pinkbeenyeti replied:
```
DROP TABLE IF EXISTS USERS;
CREATE TABLE USERS (
    id BIGINT AUTO_INCREMENT PRIMARY KEY,
    email VARCHAR(255) NOT NULL UNIQUE,
    password VARCHAR(255) NOT NULL
    role VARCHAR(50) NOT NULL,
    CHECK (role IN ('user', 'admin', 'collaborator'))
);
```
회원의 역할을 추가하기 위해 role이라는 colum을 추가하고 일반 사용자, 관리자, 관리자에게 권한 허가를 받은 협력자로 역할을 구분하기 위해 제한을 두었습니다!

dc-choi commented on src/main/java/gift/controller/ProductController.java:
```diff
@@ -0,0 +1,97 @@
+
+    @GetMapping("v3/products")
+    public String getAllProducts(Model model) {
+        List<Product> products = getProducts();
```
DTO에 대한 고민이 필요해보입니다. 힌트를 드리자면 일반적으로 Request와 Response에 대한 DTO를 따로따로 선언합니다.

pinkbeenyeti replied:
ProductService 계층을 통해 DTO를 받을 수 있도록 하였습니다!

```
    @GetMapping
    public String getAllProducts(Model model) {
        ProductsDTO products = productService.getAllProducts();
        model.addAttribute("products", products);
        retu
```rn "manage";
    }
```
ProductsDTO 객체입니다!

```
public record ProductsDTO(List<Product> productDTOList) {
}
```

pinkbeenyeti replied:
또한 제가 이해한 것이 맞는 지 궁금합니다. 예를 들어 ProductRequestDTO와 ProductResponseDTO로 나누는 이유는 요청을 받을 때는 각 필드에 대한 검증이 필요하지만 응답을 할 때는 DB에서 받아오므로 검증을 굳이 할 필요가 없어 사용하는 것인 지 여쭤보고 싶습니다!

dc-choi commented on src/main/java/gift/exceptions/KakaoContainException.java:
```diff
@@ -0,0 +1,7 @@
+package gift.exceptions;
+
+public class KakaoContainException extends RuntimeException{
```
이 객체는 카카오라는 이름을 사용했을 경우에만 예외 처리가 가능해보이는 네이밍입니다. 좀 더 포괄적인 이름으로 변경해주시고 그 안의 메시지에서 어떤 오류를 발생시키는지에 대한 처리가 필요해보입니다.

pinkbeenyeti replied:
```
package gift.exceptions;

public class InvalidNameException extends RuntimeException{
    public InvalidNameException(String message) {
        super(message);
    }
}
```

위와 같이 유효하지 않은 이름일 경우 예외를 출력하는 클래스임을 나타내도록 변경하였습니다!

dc-choi commented on src/main/java/gift/repository/UserRepository.java:
```diff
@@ -0,0 +1,44 @@
+
+    public Optional<Integer> countUsers(String email) {
+        String sql = "select count(*) from USERS where email = ?";
+        Integer count = jdbcTemplate.queryForObject(sql, new Object[]{email}, Integer.class);
```
`queryForObject()`에서 0이나 1을 초과하는 행을 반환하게 되면 Exception을 발생시킵니다. try/catch로 감싸는게 필요해보입니다.

pinkbeenyeti replied:
```
public Optional<Integer> countUsers(String email) {
        try {
            Integer count = jdbcTemplate.queryForObject(userCountSql, new Object[]{email}, Integer.class);
            return Optional.ofNullable(count);
        } catch (EmptyResultDataAccessException e) {
            return Optional.of(0);
        } catch (IncorrectResultSizeDataAccessException e) {
            return Optional.empty();
        }
    }
```

수정하였습니다!

dc-choi commented on src/main/java/gift/service/AuthService.java:
```diff
@@ -0,0 +1,19 @@
+    }
+
+    public boolean comparePassword(UserDTO userDTO) {
+        UserDTO realUser = userRepository.selectUser(userDTO.email());
```
DTO와 Entity에 대한 고민이 필요해보입니다. DTO와 Entity의 변환은 어디서 이루어져야 한다고 생각하나요?

pinkbeenyeti replied:
아하 서비스에서 변환을 하는 것이 아닌 레포지토리에서 직접 변환을 하고 있습니다. 수정하겠습니다!

dc-choi commented on src/main/java/gift/repository/UserRepository.java:
```diff
@@ -0,0 +1,44 @@
+        return new UserDTO(user.getEmail(), user.getPassword());
+    }
+
+    public Optional<Integer> countUsers(String email) {
```
`Optional`을 사용하려는 시도는 좋았습니다!

pinkbeenyeti replied:
감사합니다!

dc-choi commented on src/main/java/gift/exceptions/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,43 @@
+    public ResponseEntity<Map<String, String>> handleInvalidNameExceptions(MethodArgumentNotValidException ex) {
+        Map<String, String> errors = new HashMap<>();
+
+        ex.getBindingResult().getAllErrors().forEach(error -> {
```
이 부분에서 첫번째 오류메세지를 꺼내서 사용할 수 있습니다. 조금 더 고민해주세요.

pinkbeenyeti replied:
```
 @ExceptionHandler(MethodArgumentNotValidException.class)
    public ResponseEntity<Map<String, String>> handleInvalidNameExceptions(MethodArgumentNotValidException ex) {
        Map<String, String> error = new HashMap<>();

        String errorMessage = ex.getBindingResult()
                .getAllErrors()
                .get(0)
                .getDefaultMessage();

        error.put("message", errorMessage);
        return new ResponseEntity<>(error, HttpStatus.BAD_REQUEST);
    }
```

첫 번째 객체를 가져와 ObjectError 타입의 메서드인 getDefaultMessage()를 사용하여 원하던 에러 메시지를 가져와 응답하는 식으로 변경하였습니다!

dc-choi commented on src/main/java/gift/controller/LoginController.java:
```diff
@@ -0,0 +1,61 @@
+    @Autowired
+    private AuthService authService;
+
+    private final String emailErrorMessage = "존재하지 않는 이메일입니다.";
```
에러 메세지는 공통으로도 사용할 수 있어서 따로 객체를 만들어서 빼두는 걸 추천드립니다.

pinkbeenyeti replied:
package gift.common;

public class ErrorMessage {
    private ErrorMessage(){}

    public static final String emailAlreadyExists = "이미 가입한 이메일입니다.";
    public static final String emailInvalid = "존재하지 않는 이메일입니다.";
    public static final String passwordInvalid = "잘못된 비밀번호입니다.";

}

위와 같이 객체를 따로 만들고 공통으로 사용할 수 있도록 static으로 지정하였습니다. 또한 생성자의 접근 제어자를 private으로 할당하여  외부에서 인스턴스를 만들지 않고 사용하도록 강제하였습니다!

dc-choi commented on src/main/java/gift/controller/RegisterController.java:
```diff
@@ -0,0 +1,50 @@
+    @Autowired
+    private UserService userService;
+
+    private final String erroeMessage = "이미 가입한 이메일입니다.";
```
에러 메세지는 공통으로도 사용할 수 있어서 따로 객체를 만들어서 빼두는 걸 추천드립니다.

pinkbeenyeti replied:
위와 같습니다! 감사합니다!

pinkbeenyeti committed:
fix: 코드의 가독성을 높이기 위한 수정

dc-choi reviewed (changes requested):
안녕하세요 규현님~ PR 리뷰를 확인하였습니다~

자세한 내용은 코멘트에 남겨두었습니다~! 확인 부탁드립니다.

질문을 남겨주셨는데 그에 대한 답변을 드리도록 하겠습니다!

> 계층의 분리가 올바르게 된 것인지 궁금합니다. 기존의 코드에서 제가 부족했던 점이 무엇이었는 지 생각을 해보니 컨트롤러에서 너무 많은 기능을 수행하고 있다고 느껴 각 기능 별로 분리하는 것에 초점을 맞추었습니다. 이전의 코드는 시간 부족으로 이번 과제처럼 아직 분리하지는 못했지만 빠른 시일 내에 할 예정입니다.

각 레이어의 역할이 무엇인지 고민해보시고 학습해주세요. 이 부분에 대한 내용은 코멘트에도 작성하였습니다!!

> 통일성 이전의 코드는 특히 클라이언트로 데이터를 보내는 방식이 너무 통일성이 없다고 생각했습니다. 이번에 일관성을 지키기 위해 노력했는데 잘 된 것인지 궁금합니다.

DTO를 나누는 기준에 대해서 고민을 해볼 필요가 있어보입니다. 각 EndPoint마다 DTO를 따로 둘 것이냐 하나의 ProductDTO를 두고 Product에 대한 요청은 전부 ProductDTO로 둘 것인지에 대한 고민이 필요해보입니다.

어려웠던 점에 대해서도 남겨주셨는데 그에 대한 답도 드리겠습니다!

> Basic 인증의 적용이 어려웠습니다. Basic 인증이 원활하지 않아 db에서 값을 가져 와 직접 비교하는 식으로 구현하였습니다.

어떤 부분에서 DB에서 값을 가져와 직접 비교하는지 코드로 찾을 수 없었습니다. Spring Security에 대해 좀 더 학습하실 필요가 있어보이십니다.

dc-choi commented on src/main/java/gift/repository/UserRepository.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    public Optional<Integer> countUsers(String email) {
+        Integer count = jdbcTemplate.queryForObject(userCountSql, new Object[]{email}, Integer.class);
```
`queryForObject()`에서 0이나 1을 초과하는 행을 반환하게 되면 Exception을 발생시킵니다. try/catch로 감싸는게 필요해보입니다.

pinkbeenyeti committed:
fix: 스키마 수정 및 데이터 삽입
USERS 테이블에서 권한을 주기 위해 role 컬럼을 추가하고 user, admin,
collaborator의 값만 가질 수 있도록 제약을 걸었다.

pinkbeenyeti committed:
feat: ProductsDTO 추가
여러 상품에 대한 데이터를 전송하기 위한 객체를 생성하였다.

pinkbeenyeti committed:
feat: ProductRepository 생성
컨트톨러가 하는 역할이 너무 많아 분리하기 위해 생성하였다.
DB와 직접적인 상호 작용을 하고 서비스와 상호작용 한다.

pinkbeenyeti committed:
fix: 컨트롤러와 서비스 계층 분리
서비스 계층을 생성하였다. 비즈니스 로직을 처리한다.

pinkbeenyeti committed:
feat: 프로덕스 서비스 계층 생성

pinkbeenyeti committed:
fix: 로그인 화면에서 입력 창을 빈칸으로 두지 않도록 수정
그 외 별다른 수정사항 없음

pinkbeenyeti committed:
feat: 새로운 예외 객체 생성
원활한 예외 처리를 위한 예외 객체를 생성하였다.

pinkbeenyeti committed:
fix: role 추가

pinkbeenyeti committed:
feat: 에러 메시지 객체 생성
공통적으로 많이 사용하는 에러 메시지를 사용하기 위한 객체를
생성하였다ㅓ.

pinkbeenyeti committed:
feat: 토큰의 시크릿 키에 대한 보안성을 위해 속성 추가

pinkbeenyeti committed:
fix: 직접적인 에러 검출에서 예외 발생으로 변경

pinkbeenyeti committed:
fix: 추가로 생성한 예외 객체에 대한 처리하도록 수정

pinkbeenyeti committed:
fix: 안전을 위한 try catch 블록 사용

pinkbeenyeti committed:
fix: 각자의 역할에 맞는 메서드를 가지도록 수정

pinkbeenyeti committed:
feat: DB 테이블 추가 및 데이터 사전 입력

pinkbeenyeti committed:
feat: 위시리스트 페이지 구현

pinkbeenyeti committed:
fix: Product 관련 클래스 가독성 개선

pinkbeenyeti committed:
fix: 로그인 페이지, 토큰 전송 기능 추가

pinkbeenyeti committed:
fix: ProductDTO 분리

pinkbeenyeti committed:
feat: jwt 토큰 사용 사용자 구분 기능 구현

pinkbeenyeti committed:
fix: user 코드 개선

pinkbeenyeti committed:
feat: webconfig 추가

pinkbeenyeti committed:
feat: 예외 객체 추가

pinkbeenyeti committed:
feat: 위시리스트 기능 구현

pinkbeenyeti committed:
feat: 위시리스트 품목 추가 페이지

pinkbeenyeti committed:
feat: 코드 개선

pinkbeenyeti committed:
feat: 별다른 개선 사항 없음

dc-choi commented on src/main/java/gift/annotation/LoginUserArgumentResolver.java:
```diff
@@ -0,0 +1,65 @@
+
+    @Override
+    public boolean supportsParameter(MethodParameter parameter) {
+        return parameter.hasParameterAnnotation(LoginUser.class) && parameter.getParameterType().equals(User.class);
```
`parameter.getParameterType().equals(User.class)`를 아래처럼 변경해주세요.

`User.class.isAssignableFrom(parameter.getParameterType())`

pinkbeenyeti replied:
변경 완료했습니다!

dc-choi commented on src/main/java/gift/exceptions/InvalidNameException.java:
```diff
@@ -0,0 +1,7 @@
+package gift.exceptions;
+
+public class InvalidNameException extends RuntimeException{
```
`InvalidNameException`, `InvalidUserException`, `UserNotFoundException`을 공통의 객체를 사용해서 처리하는건 어떻게 생각하시나요?

`CustomException`으로 처리를 하고 그 안에 메시지로 구분을 하는 것이 유지보수성에 더 좋아보입니다~!

물론 이건 나중에 하셔도 됩니다.

pinkbeenyeti replied:
예외 상황이 많아질 수록 예외 객체가 많아지고, GlobalHandler에서 처리해야 할 예외 객체 또한 많아지기 때문에 그러한 편이 유지보수성에 좋을 것 같습니다! 한 가지 질문사항이 있습니다.

` CustomException`으로 처리를 하고 그 안에 메시지로 구분을 할 때 메시지마다 보내려는 HTTP 상태 코드가 다른 상황인데 GlobalHandler가 이 메시지에 대한 구분을 어떻게 하면 좋을까요?

dc-choi replied:
예외가 발생할 때 HTTP status 값을 보내줘서 `GlobalHandler`가 그 값을 처리하는 방식이면 될 거 같습니다~!

pinkbeenyeti replied:
아하! 이해했습니다 감사합니다!

dc-choi reviewed (changes requested):
규현님~! 정말 고생하셨어요... 전주랑 다른 사람이 코딩한 수준으로 완전 열심히 하셨네요...

코멘트 달아주신 부분은 아래 하단 코멘트에 적어두었습니다~!

그 부분만 수정되면 바로 승인하도록 할게요~!

pinkbeenyeti committed:
fiz: resolver 코드 개선

pinkbeenyeti committed:
feat: 예외 객체 유지 보수를 위한 커스텀 예외 객체 생성

dc-choi reviewed (approved):
규현님~! step3까지 과제 구현하시느라 고생하셨습니다~!

수고하셨습니다! 3주차도 파이팅~!
